shotover-proxy/src/transforms/protect/encrypt.rs

```diff
@@ -0,0 +1,66 @@
+use crate::message::MessageValue;
+use crate::transforms::protect::key_management::KeyManager;
+use anyhow::{anyhow, bail, Result};
+use serde::{Deserialize, Serialize};
+use sodiumoxide::crypto::secretbox;
+use sodiumoxide::crypto::secretbox::Nonce;
+use sqlparser::ast::Value as SQLValue;
+
+// A protected value meets the following properties:
+// https://doc.libsodium.org/secret-key_cryptography/secretbox
+// This all relies on crypto_secretbox_easy which takes care of
+// all padding, copying and timing issues associated with crypto
+#[derive(Serialize, Deserialize)]
+struct Protected {
+    cipher: Vec<u8>,
+    nonce: Nonce,
+    enc_dek: Vec<u8>,
+    kek_id: String,
+}
+
+pub async fn encrypt(
+    value: &SQLValue,
+    key_management: &KeyManager,
+    key_id: &str,
+) -> Result<SQLValue> {
+    let value = MessageValue::from(&*value);
+
+    let sym_key = key_management.cached_get_key(key_id, None, None).await?;
+
+    let ser = bincode::serialize(&value)?;
```

Comment:
as with decrypt should we customise this error message back to the client?

Reply:
nah, I think we want to return as much information as possible when handling a a serialization error.
An error like this should never ever occur so if it does we will need as much information as possible to diagnose it.

The value is already encrypted at this point so no risk of leaking data by directly returning whatever the serialization error contains.